topi/python/topi/cuda/conv2d_nhwc_tensorcore.py

```diff
@@ -0,0 +1,361 @@
+    # compute the output shape
+    dilated_kernel_h = (kernel_h - 1) * dilation_h + 1
+    dilated_kernel_w = (kernel_w - 1) * dilation_w + 1
+    pad_top, pad_left, pad_down, pad_right = get_pad_tuple(
+        padding, (dilated_kernel_h, dilated_kernel_w))
+    out_channel = num_filter
+    out_height = simplify((in_height - dilated_kernel_h + pad_top + pad_down) // stride_h + 1)
+    out_width = simplify((in_width - dilated_kernel_w + pad_left + pad_right) // stride_w + 1)
+    pad_before = [0, pad_top, pad_left, 0]
+    pad_after = [0, pad_down, pad_right, 0]
+    PaddedInput = pad(Input, pad_before, pad_after, name="PaddedInput")
+    rc = te.reduce_axis((0, in_channel), name='rc')
+    ry = te.reduce_axis((0, kernel_h), name='ry')
+    rx = te.reduce_axis((0, kernel_w), name='rx')
+    # convert data type of input feature maps and weights
+    TransPaddedInput = te.compute(
+        PaddedInput.shape,
+        lambda h, w, i, o: PaddedInput[h, w, i, o].astype('float16'))
+    TransFilter = te.compute(
+        Filter.shape, lambda h, w, i, o: Filter[h, w, i, o].astype('float16'))
+    Output = te.compute(
+        (batch, out_height, out_width, out_channel),
+        lambda nn, yy, xx, ff: te.sum(
+            TransPaddedInput[nn, yy * stride_h + ry * dilation_h,
+                             xx * stride_w + rx * dilation_w, rc].astype(out_dtype) *
+            TransFilter[ry, rx, rc, ff].astype(out_dtype), axis=[ry, rx, rc]),
+        name="Conv2dOutput", tag="conv2d_nhwc_tensorcore")
+    return Output
+
+
+def schedule_nhwc_tensorcore_cuda(cfg, s, Conv):
+    """Schedule tensorcore template"""
+    kh, kw, ic = s[Conv].op.reduce_axis
+    out_dtype = Conv.dtype
+    trans_paddata, kernel = s[Conv].op.input_tensors
+    batch, _, _, _ = get_const_tuple(Conv.shape)
+    _, _, _, out_channels = get_const_tuple(kernel.shape)
+    paddata = s[trans_paddata].op.input_tensors
+
+    # inline the pad and dtype transform
+    s[trans_paddata].compute_inline()
+    s[kernel].compute_inline()
+    s[paddata[0]].compute_inline()
+
+    # Designate the memory hierarchy
+    if Conv.op in s.outputs:
+        output = Conv
+        kh, kw, ic = s[Conv].op.reduce_axis
+        AS = s.cache_read(trans_paddata, 'shared', [Conv])
+        WS = s.cache_read(kernel, 'shared', [Conv])
+        AF = s.cache_read(AS, 'wmma.matrix_a', [Conv])
+        WF = s.cache_read(WS, 'wmma.matrix_b', [Conv])
+        ConvF = s.cache_write(Conv, 'wmma.accumulator')
+        ConvS = s.cache_read(ConvF, 'shared', [Conv])
+        OL = ConvS
+    else:
+        output = s.outputs[0].output(0)
+        s[Conv].set_scope('shared')
+        OL = Conv
+        kh, kw, ic = s[OL].op.reduce_axis
+        AS = s.cache_read(trans_paddata, 'shared', [OL])
+        WS = s.cache_read(kernel, 'shared', [OL])
+        AF = s.cache_read(AS, 'wmma.matrix_a', [OL])
+        WF = s.cache_read(WS, 'wmma.matrix_b', [OL])
+        ConvF = s.cache_write(OL, 'wmma.accumulator')
```

Comment:
this if/else has a lot of duplicated code that could be shared.

Reply:
The duplicated codes have been placed before the if/else block.

Reply:
The duplicated codes have been placed before the if/else block.